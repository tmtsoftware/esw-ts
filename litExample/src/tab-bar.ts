import { customElement, state } from 'lit/decorators.js'
import { LitElement, html } from 'lit'
import '@vaadin/tabs'

@customElement('tab-bar')
export class TabBar extends LitElement {

  render() {
    return html`
      <vaadin-tabs>
        <vaadin-tab>Public</vaadin-tab>
        <vaadin-tab>Secured</vaadin-tab>
        <vaadin-tab>Config App</vaadin-tab>
        <vaadin-tab>Admin App</vaadin-tab>
        <vaadin-tab>User App</vaadin-tab>
      </vaadin-tabs>
    `
  }
}
