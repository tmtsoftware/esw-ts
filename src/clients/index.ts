export * from './aas'
export * from './admin'
export * from './agent-service'
export * from './alarm'
export * from './command'
export * from './config-service'
export * from './event'
export * from './location'
export * from './logger'
export * from './sequence-manager'
export * from './sequencer'
