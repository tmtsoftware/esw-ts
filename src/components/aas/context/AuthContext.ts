import { createContext } from 'react'
import type { Auth } from '../../../clients/aas'

export interface AuthContextType {
  auth: Auth | null
  login: () => void
  logout: () => void
}

const AuthContextDefaultState = {
  auth: null,
  login: () => undefined,
  logout: () => undefined
}
/**
 * Default state for AuthContextProvider
 * @type {{auth: undefined, login: (function(): boolean), logout: (function(): boolean)}}
 */
const AuthContext = createContext<AuthContextType>(AuthContextDefaultState)
const { Provider, Consumer } = AuthContext

// todo: AuthContext is exported to support scala.js, see if only exporting Consumer works
export { AuthContext, Provider, Consumer, AuthContextDefaultState }
