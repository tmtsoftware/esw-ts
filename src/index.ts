export * from './clients/aas'
export * from './clients/admin'
export * from './clients/agent-service'
export * from './clients/alarm'
export * from './clients/command'
export * from './clients/config-service'
export * from './clients/event'
export * from './clients/location'
export * from './clients/logger'
export * from './clients/sequence-manager'
export * from './clients/sequencer'
export * from './components'
export * from './config'
export * from './models'
