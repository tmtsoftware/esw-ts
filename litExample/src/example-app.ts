import { Auth, loadGlobalConfig } from '@tmtsoftware/esw-ts'
import { LitElement, html, css } from 'lit'
import { customElement, property } from 'lit/decorators.js'
import { Task } from '@lit/task'
import litLogo from './assets/lit.svg'
import viteLogo from '/vite.svg'
import '../../src/components/aas/context/AuthContextProviderLit.ts'
import './tab-bar.ts'

/**
 * This is the top level example application element
 */
@customElement('example-app')
export class ExampleApp extends LitElement {
  @property()
  docsHint = 'Click on the Vite and Lit logos to learn more'

  @property({ type: Number })
  count = 0

  // XXX TODO FIXME
  @property({ attribute: false, reflect: true })
  auth: Auth | null = null

  @property({ type: Boolean })
  failed = false

  private initTask = new Task(this, {
    task: async () => {
      return await loadGlobalConfig().catch(() => {
        console.log('Config.js not found. Failed to Load Global Configuration.')
      })
    },
    args: () => []
  })

  render() {
    return this.initTask.render({
      initial: () => html`<p>Not running yet ...</p>`,
      pending: () => html`<p>Loading ...</p>`,
      error: (e) => html`<p>Error: ${e}</p>`,
      complete: () => this.renderMain()
    })
  }

  renderMain() {
    return html`
      <tab-bar></tab-bar>
      <auth-context-provider realm="TMT" client-id="tmt-frontend-app">
        <div>
          <a href="https://vitejs.dev" target="_blank">
            <img src=${viteLogo} class="logo" alt="Vite logo" />
          </a>
          <a href="https://lit.dev" target="_blank">
            <img src=${litLogo} class="logo lit" alt="Lit logo" />
          </a>
        </div>
        <slot></slot>
        <div class="card">
          <button @click=${this._onClick} part="button">count is ${this.count}</button>
        </div>
        <p class="read-the-docs">${this.docsHint}</p>
      </auth-context-provider>
    `
  }

  private _onClick() {
    this.count++
  }

  static styles = css`
    :host {
      max-width: 1280px;
      margin: 0 auto;
      padding: 2rem;
      text-align: center;
    }

    .logo {
      height: 6em;
      padding: 1.5em;
      will-change: filter;
      transition: filter 300ms;
    }

    .logo:hover {
      filter: drop-shadow(0 0 2em #646cffaa);
    }

    .logo.lit:hover {
      filter: drop-shadow(0 0 2em #325cffaa);
    }

    .card {
      padding: 2em;
    }

    .read-the-docs {
      color: #888;
    }

    ::slotted(h1) {
      font-size: 3.2em;
      line-height: 1.1;
    }

    a {
      font-weight: 500;
      color: #646cff;
      text-decoration: inherit;
    }

    a:hover {
      color: #535bf2;
    }

    button {
      border-radius: 8px;
      border: 1px solid transparent;
      padding: 0.6em 1.2em;
      font-size: 1em;
      font-weight: 500;
      font-family: inherit;
      background-color: #1a1a1a;
      cursor: pointer;
      transition: border-color 0.25s;
    }

    button:hover {
      border-color: #646cff;
    }

    button:focus,
    button:focus-visible {
      outline: 4px auto -webkit-focus-ring-color;
    }

    @media (prefers-color-scheme: light) {
      a:hover {
        color: #747bff;
      }

      button {
        background-color: #f9f9f9;
      }
    }
  `
}

declare global {
  interface HTMLElementTagNameMap {
    'example-app': ExampleApp
  }
}
