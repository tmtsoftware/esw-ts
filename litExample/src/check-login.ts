import { consume } from '@lit/context'
import { LitElement, html } from 'lit'
import { customElement, property } from 'lit/decorators.js'
import { AuthContextLit, litAuthContext } from '@tmtsoftware/esw-ts'

@customElement('check-login')
export class CheckLogin extends LitElement {
  @consume({ context: litAuthContext })
  @property({ attribute: false })
  private authContext?: AuthContextLit

  render() {
    if (this.authContext?.auth && this.authContext.auth.isAuthenticated()) {
      return html`<div><slot name="ok"></slot></div>`
    } else {
      return html`<div><slot name="error"></slot></div>`
    }
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'check-login': CheckLogin
  }
}
