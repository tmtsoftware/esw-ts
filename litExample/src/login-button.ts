import { consume } from '@lit/context'
import { LitElement, html } from 'lit'
import { customElement, property } from 'lit/decorators.js'
import { AuthContextLit, litAuthContext, setAppName } from '@tmtsoftware/esw-ts'

setAppName('example')

@customElement('login-button')
export class LoginButton extends LitElement {
  @consume({ context: litAuthContext, subscribe: true })
  @property({ attribute: false })
  private authContext?: AuthContextLit

  private async handleLogin() {
    await this.authContext?.login()
  }

  private async handleLogout() {
    await this.authContext?.logout()
  }

  render() {
    if (this.authContext?.auth && this.authContext.auth.isAuthenticated()) {
      return html` <sl-button @click=${this.handleLogout}>Logout</sl-button>`
    } else {
      return html`<sl-button @click=${this.handleLogin}>Login</sl-button>`
    }
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'login-button': LoginButton
  }
}
