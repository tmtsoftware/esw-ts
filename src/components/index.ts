import CheckLogin from './aas/authentication/CheckLogin'
import RealmRole from './aas/authorization/RealmRole'
import { AuthContext, Consumer, AuthContextType } from './aas/context/AuthContext'
import AuthContextProvider from './aas/context/AuthContextProvider'
import Login from './aas/Login'
import Logout from './aas/Logout'

// #export-components
export { Logout, Login, CheckLogin, RealmRole, AuthContextProvider, Consumer, AuthContext }

export type { AuthContextType }
// #export-components
