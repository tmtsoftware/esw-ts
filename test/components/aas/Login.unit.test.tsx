import { fireEvent, render } from '@testing-library/react'
import React from 'react'
import { Login } from '../../../src/components'
import { Provider } from '../../../src/components/aas/context/AuthContext'
import { mockAuth } from '../../helpers/MockHelpers'

// DEOPSCSW-630 - Javascript adapter for AAS
// DEOPSCSW-631 - React layer for javascript adapter for AAS
describe('<Login />', () => {
  test('should call login | ESW-330', async () => {
    const mockLogin = jest.fn()
    const { getByText } = await render(
      <Provider value={{ auth: mockAuth(), login: mockLogin, logout: jest.fn() }}>
        <Login />
      </Provider>
    )

    fireEvent.click(getByText('Login'))

    expect(mockLogin).toHaveBeenCalled()
  })

  test('should render login | ESW-330', async () => {
    const { container } = render(<Login />)
    expect(container).toMatchSnapshot()
  })
})
