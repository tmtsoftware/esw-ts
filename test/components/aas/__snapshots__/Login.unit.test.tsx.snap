// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should render login | ESW-330 1`] = `
<div>
  <button
    id="keycloak-login"
  >
    Login
  </button>
</div>
`;
